### PR TITLE
Interpolate vagrantfile_template on Vagrant post-processor

### PR DESCRIPTION
**Pull request by sylviamoss**
This is to allow using template variables in the Vagrantfile template.

Closes https://github.com/hashicorp/packer/issues/9921

### CODE REVIEW

**Review by nywilken — COMMENTED** [review]
This looks good to me. Do we need to update documentation to indicate that template variables can be used with the vagrantfile?

File: builder/docker/driver.go [hunk 1]
```diff
@@ -26,6 +26,9 @@ type Driver interface {
 	// for external access.
 	IPAddress(id string) (string, error)
 
+	// Sha256 returns the sha256 id of the image
+	Sha256(id string) (string, error)
```
[the remark this answers is not in the document]
Comment (sylviamoss) [reply builder/docker/driver.go:30]: No, is not the same value! The container ID is different from the Image id when the container is committed to one.

[the remark this answers is not in the document]
Comment (sylviamoss) [reply builder/docker/driver.go:30]:
```
    docker.base_setup: Image ID: sha256:f70456b832a676702ca45cc80d909196f939ba793a8dd476c7bb3bed78250f63
==> docker.base_setup: Killing the container: daf68c7f441da7c4e0f397419b10823662ac829cbffcc5e1f12b4b4425cd4c67

```

[the remark this answers is not in the document]
Comment (nywilken) [reply builder/docker/driver.go:30]: Ah yes, I see. Thank you.

File: builder/docker/builder.go [hunk 2]
```diff
@@ -67,23 +80,18 @@ func (b *Builder) Run(ctx context.Context, ui packer.Ui, hook packer.Hook) (pack
 		log.Print("[DEBUG] Container will be discarded")
 	} else if b.config.Commit {
 		log.Print("[DEBUG] Container will be committed")
-		steps = append(steps, new(StepCommit))
+		steps = append(steps,
+			new(StepCommit),
+			&StepSetGeneratedData{ // Adds ImageSha256 variable available after StepCommit
```
Comment (sylviamoss) [inline builder/docker/builder.go:85]: @nywilken updated the code to make it easier to see when the variable the available

File: website/pages/docs/builders/docker.mdx [hunk 3]
```diff
@@ -215,6 +215,16 @@ You must specify (only) one of `commit`, `discard`, or `export_path`.
...
 @include 'builder/docker/Config-not-required.mdx'
 
+## Build Shared Information Variables
+
+This builder generates data that are shared with provisioner and post-processor via build function of [template engine](/docs/templates/engine)
+for JSON and [contextual variables](/docs/from-1.5/contextual-variables) for HCL2.
+
+The generated variable available for this builder is:
+
+- `ImageSha256` - When committing a container to an image, this will give the image SHA256. Because the image is not available at the provision step,
```
Comment (sylviamoss) [inline website/pages/docs/builders/docker.mdx:225]: @nywilken and also the docs 👍

File: website/pages/docs/builders/docker.mdx [hunk 4]
```diff
@@ -215,6 +215,16 @@ You must specify (only) one of `commit`, `discard`, or `export_path`.
 
 @include 'builder/docker/Config-not-required.mdx'
 
+## Build Shared Information Variables
+
+This builder generates data that are shared with provisioner and post-processor via build function of [template engine](/docs/templates/engine)
```
Comment (SwampDragons) [inline website/pages/docs/builders/docker.mdx:220]:
```suggestion
This build shares generated data with provisioners and post-processors via [template engines](/docs/templates/engine)
```

File: website/pages/docs/builders/docker.mdx [hunk 5]
```diff
@@ -215,6 +215,16 @@ You must specify (only) one of `commit`, `discard`, or `export_path`.
...
+for JSON and [contextual variables](/docs/from-1.5/contextual-variables) for HCL2.
+
+The generated variable available for this builder is:
+
+- `ImageSha256` - When committing a container to an image, this will give the image SHA256. Because the image is not available at the provision step,
+this variable is only available for post-processor.
```
Comment (SwampDragons) [inline website/pages/docs/builders/docker.mdx:226]:
```suggestion
this variable is only available for post-processors.
```

**Review by SwampDragons — APPROVED** [review]

**Pull request merged**